File: app/src/main/java/com/worldventures/dreamtrips/social/ui/profile/service/command/GetPublicProfileCommand.java
package com.worldventures.dreamtrips.social.ui.profile.service.command;

import com.worldventures.core.janet.CommandWithError;
import com.worldventures.janet.injection.InjectableAction;
import com.worldventures.core.model.User;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.profile.GetPublicUserProfileHttpAction;

import javax.inject.Inject;

import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;
import io.techery.mappery.MapperyContext;

@CommandAction
public class GetPublicProfileCommand extends CommandWithError<User> implements InjectableAction {

   @Inject Janet janet;
   @Inject MapperyContext mappery;

   private int userId;

   public GetPublicProfileCommand(int userId) {
      this.userId = userId;
   }

   @Override
   protected void run(CommandCallback callback) throws Throwable {
      janet.createPipe(GetPublicUserProfileHttpAction.class)
            .createObservableResult(new GetPublicUserProfileHttpAction(userId))
            .map(GetPublicUserProfileHttpAction::response)
            .map(user -> mappery.convert(user, User.class))
            .subscribe(callback::onSuccess, callback::onFail);
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.error_fail_to_load_profile_info;
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/profile/command/GetPublicProfileCommand.kt
package com.worldventures.dreamtrips.social.service.profile.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.core.model.User
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.profile.GetPublicUserProfileHttpAction
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import io.techery.mappery.MapperyContext
import javax.inject.Inject

@CommandAction
class GetPublicProfileCommand(private val userId: Int) : CommandWithError<User>(), InjectableAction {

   @Inject lateinit var janet: Janet
   @Inject lateinit var mappery: MapperyContext

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<User>) {
      janet.createPipe(GetPublicUserProfileHttpAction::class.java)
            .createObservableResult(GetPublicUserProfileHttpAction(userId))
            .map(GetPublicUserProfileHttpAction::response)
            .map { mappery.convert(it, User::class.java) }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   override fun getFallbackErrorMessage() = R.string.error_fail_to_load_profile_info

}
